File: vfr-green-map/src/resources/data-provider.ts
import * as $ from 'jquery'
import { CachedProvider, CachedProviderRequest } from './cached-provider'
import { Point2d, PointWebMercator, PointGeo, CoordinateConversion, BoxGeo } from "coordinates"
import { IDataReceiver } from "./i-data-receiver";
import { OverlayTypes } from '../view/overlay-types'

export class DataProvider extends CachedProvider {
    private static readonly AREA_BUCKET_MULTIPLIER = 2;
    private static readonly REQUEST_DELAY_MILLISECONDS = 700;
    private usedBuckets: Record<OverlayTypes, Record<string, Array<BoxGeo>>>; //<overlay_type, <resolution as string, box geos>>

    constructor() {
        super(DataProvider.REQUEST_DELAY_MILLISECONDS);
        this.usedBuckets = {
            0: {},
            1: {},
            2: {}, 
            3: {},
            4: {},
            5: {},
            6: {},
            7: {}
        };
    }

    public retrieveTile(area: BoxGeo, resolution: PointGeo, type: OverlayTypes, receiver: IDataReceiver): void {
        let resolutionBucket = this.createResolutionBucket(resolution);
        let areaBucket = this.createAreaBucket(area, resolutionBucket, type);
        let key = this.getCacheKey(areaBucket, resolutionBucket, type);
        let tileRequest = this.getExistingRequest(key) as DataRequest;

        if ( tileRequest !== undefined && !tileRequest.isInError() && !tileRequest.isExpired() ) {
            let tileRequestScoped = tileRequest as DataRequest;
            if (tileRequest.isLoaded()) {
                tileRequestScoped.setReceiver(receiver);
                tileRequestScoped.broadcastData(true);
            } else {
                tileRequestScoped.setReceiver(receiver);
                //this.findTemporaryData(mapName, zoomLevel, location, receiver, data);
            }
        } else {
            let newRequest = new DataRequest(this, receiver, areaBucket, resolutionBucket, type);
            this.addRequestToQueue(key, newRequest);
            //this.findTemporaryData(mapName, zoomLevel, location, receiver, data);
        }
    }

    private createResolutionBucket(resolution: PointGeo): PointGeo {
        let x = Math.log(resolution.longitude) / Math.log(2)
        x = Math.ceil(x / 2) * 2
        let y = Math.log(resolution.latitude) / Math.log(2)
        y = Math.ceil(y / 2) * 2

        return new PointGeo(2 ** x, 2 ** y);
    }

    private createAreaBucket(area: BoxGeo, resolution: PointGeo, type: OverlayTypes): BoxGeo {
        // Check if the area exists in a used bucket
        let resolutionKey = this.getResolutionKey(resolution);
        if ( !(resolutionKey in this.usedBuckets[type]) ) {
            this.usedBuckets[type][resolutionKey] = new Array<BoxGeo>();
        }
        let areaTl = area.getTopLeft();
        let areaBr = area.getBottomRight();
        for ( let obj of this.usedBuckets[type][resolutionKey] ) {
            let tl = obj.getTopLeft();
            let br = obj.getBottomRight();
            if ( tl.longitude <= areaTl.longitude && tl.latitude >= areaTl.latitude &&
                br.longitude >= areaBr.longitude && br.latitude <= areaBr.latitude ) {
                return obj;
            }
        }

        // We need to create a new bucket
        let areaDimensions = area.getDimensions();
        let tlLongitude = areaTl.longitude - (areaDimensions.longitude * DataProvider.AREA_BUCKET_MULTIPLIER);
        let tlLatitude = areaTl.latitude + (areaDimensions.latitude * DataProvider.AREA_BUCKET_MULTIPLIER);
        let brLongitude = areaBr.longitude + (areaDimensions.longitude * DataProvider.AREA_BUCKET_MULTIPLIER);
        let brLatitude = areaBr.latitude - (areaDimensions.latitude * DataProvider.AREA_BUCKET_MULTIPLIER);
        let bucket = new BoxGeo(new PointGeo(this.forceValidLongitude(tlLongitude), this.forceValidLatitude(tlLatitude)),
            new PointGeo(this.forceValidLongitude(brLongitude), this.forceValidLatitude(brLatitude)));
        this.usedBuckets[type][resolutionKey].push(bucket);
        return bucket;
    }

    private forceValidLongitude(longitude: number): number {
        if ( longitude < -180 ) {
            return -180;
        }
        if ( longitude > 180 ) {
            return 180;
        }
        return longitude;
    }

    private forceValidLatitude(latitude: number): number {
        if ( latitude < -90 ) {
            return -90;
        }
        if ( latitude > 90 ) {
            return 90;
        }
        return latitude;
    }

    private getResolutionKey(resolution: PointGeo): string {
        return `${resolution.longitude}|${resolution.latitude}`;
    }

    private getCacheKey(area: BoxGeo, resolution: PointGeo, type: OverlayTypes): string {
        let tl = area.getTopLeft();
        let br = area.getBottomRight();
        return `${type}|${tl.longitude}|${tl.latitude}|${br.longitude}|${br.latitude}|${resolution.longitude}|${resolution.latitude}`
    }
}

class DataRequest extends CachedProviderRequest {
    private data: any | undefined;
    private timeReceived: number;
    private oldestDataAgeAtRetrievalSeconds: number | undefined;
    private receiver: IDataReceiver | undefined;
    private area: BoxGeo;
    private resolution: PointGeo;
    private type: OverlayTypes;
    private static readonly URL_BASE = "https://api.blueskycharts.com/condition-v2/getConditions";
    private static readonly MAX_AGE_MILLISECONDS = 5 * 60 * 1000;

    constructor(provider: DataProvider, receiver: IDataReceiver, area: BoxGeo, resolution: PointGeo, type: OverlayTypes) {
        super(provider);
        this.receiver = receiver;
        this.area = area;
        this.resolution = resolution;
        this.type = type;
        this.timeReceived = 0;
    }

    public isExpired(): boolean {
        if ( this.isLoaded() ) {
            let now = (new Date()).getTime();
            return now - this.timeReceived > DataRequest.MAX_AGE_MILLISECONDS;
        } else {
            return false;
        }
    }
    
    public sendRequest(): void {
        let tl = this.area.getTopLeft();
        let br = this.area.getBottomRight();
        let url = `${DataRequest.URL_BASE}?startLongitude=${tl.longitude}&endLongitude=${br.longitude}&startLatitude=${br.latitude}&endLatitude=${tl.latitude}&bufferLongitude=${this.resolution.longitude}&bufferLatitude=${this.resolution.latitude}&information=${this.getInformationForType(this.type)}`;
        $.ajax({
            url: url,
            type: "GET",
            dataType: "json",
            crossDomain: true,
            cache: false,
            success: (data) => {
                try {
                    this.data = data;
                    this.timeReceived = (new Date()).getTime();
                    if ( this.data !== undefined && this.data.oldestDataAgeSeconds !== undefined ) {
                        this.oldestDataAgeAtRetrievalSeconds = this.data.oldestDataAgeSeconds
                    }
                    this.completeRequest(true);
                } catch {
                    this.completeRequest(false);
                    console.error(`Could not parse data from ${url}`);
                }
            },
            error: (response) => {
                this.completeRequest(false);
                console.error(`Could not load data from ${url}`);
            }});
    }

    public setReceiver(receiver: IDataReceiver) {
        this.receiver = receiver;
    }

    public broadcastData(immediate: boolean) {
        if ( this.data === undefined || this.oldestDataAgeAtRetrievalSeconds === undefined || this.data.conditions === undefined || 
            this.receiver == undefined ) {
            return;
        }

        let secondsSinceRequest = Math.ceil(((new Date()).getTime() - this.timeReceived) / 1000);
        let dataAgeSeconds = this.oldestDataAgeAtRetrievalSeconds + secondsSinceRequest;
        for ( let data of this.data.conditions ) {
            if (data.longitude === undefined || data.latitude === undefined) {
                continue;
            }
            let geoLocation = new PointGeo(data.longitude, data.latitude);
            let mercatorLocation = CoordinateConversion.convertToWebMercator(geoLocation);

            this.receiver.receiveData(mercatorLocation, data, dataAgeSeconds, immediate);
        }
    }
    
    private getInformationForType(type: OverlayTypes): string {
        switch( type ) {
            case OverlayTypes.Category: {
                return "flightCategory"
            }
            case OverlayTypes.Ceiling: {
                return "ceiling"
            }
            case OverlayTypes.CloudCover: {
                return "cloudCover"
            }
            case OverlayTypes.DewpointSpreadC: {
                return "dewpointSpreadCelcius"
            }
            case OverlayTypes.TempC: {
                return "temperatureCelcius"
            }
            case OverlayTypes.Visibility: {
                return "visibility"
            }
            case OverlayTypes.Wind: {
                return "wind"
            }
            default: {
                console.error("Invalid overlay type");
                return "";
            }
        }
    }
}

File: vfr-green-app/app/src/main/java/com/blueskycharts/app/map/resources/DataProvider.kt
package com.blueskycharts.app.map.resources

import android.content.Context
import android.graphics.Canvas
import com.blueskycharts.app.assests.DiskCacheFactory
import com.blueskycharts.app.coordinates.BoxGeo
import com.blueskycharts.app.coordinates.PointGeo
import com.blueskycharts.app.map.view.Map
import com.blueskycharts.app.map.view.OverlayTypes
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.lang.Exception
import java.lang.Math.pow
import java.util.*
import kotlin.collections.HashMap
import kotlin.math.ceil
import kotlin.math.log2
import kotlin.math.pow

class DataProvider(val context: Context, map: Map) : CachedProvider(map, 700) {
    private val areaBucketMultiplier = 2
    private val usedBuckets: HashMap<OverlayTypes, HashMap<String, LinkedList<BoxGeo>>> = HashMap()  //<overlay_type, <resolution as string, box geos>>

    init {
        for ( type in OverlayTypes.values() ) {
            this.usedBuckets[type] = HashMap()
        }

        // Preload disk cached data
        GlobalScope.launch {
            val localFiles = DiskCacheFactory.instance.getAssetDescriptionsIn(DataRequest.endpointDescriptor)
            for (file in localFiles) {
                // If the data is expired, delete it
                if (DiskCacheFactory.instance.isExpired(file)) {
                    DiskCacheFactory.instance.deleteAsset(file)
                    continue
                }

                // Figure out what bucket the data belongs to
                val newDataRequest = DataRequest.createFromFileDescriptor(file, this@DataProvider) ?: continue
                val key = getCacheKey(newDataRequest.area, newDataRequest.resolution, newDataRequest.type)
                addDataToCache(key, newDataRequest)
                val resolutionKey = getResolutionKey(newDataRequest.resolution)
                val typeMap = usedBuckets[newDataRequest.type]
                if (typeMap != null) {
                    var listOfBuckets = typeMap[resolutionKey]
                    if (listOfBuckets == null) {
                        listOfBuckets = LinkedList()
                        typeMap[resolutionKey] = listOfBuckets
                    }
                    listOfBuckets.add(newDataRequest.area)
                }
            }
        }
    }

    fun retrieveTile(area: BoxGeo, resolution: PointGeo, type: OverlayTypes, receiver: DataReceiver, canvas: Canvas, data: Any?) {
        val resolutionBucket = this.createResolutionBucket(resolution)
        val areaBucket = this.createAreaBucket(area, resolutionBucket, type)
        val key = this.getCacheKey(areaBucket, resolutionBucket, type)
        val tileRequest = this.getCachedItem(key)
        var addNewRequest = true

        if ( tileRequest != null && tileRequest.loaded ) {
            tileRequest as DataRequest
            tileRequest.setReceiver(receiver, data)
            tileRequest.broadcastData(true, canvas)
            if ( !tileRequest.expired ) {
                addNewRequest = false
            }
        }

        if (addNewRequest) {
            incrementAwaitingQueueAddition()
            GlobalScope.launch {
                try {
                    val newTileRequest = this@DataProvider.getExistingRequest(key)
                    if (newTileRequest != null && !newTileRequest.inError &&
                        !(newTileRequest.loaded && newTileRequest.expired)) {
                        newTileRequest as DataRequest
                        newTileRequest.setReceiver(receiver, data)
                        if (newTileRequest.loaded) {
                            newTileRequest.broadcastData(false, canvas)
                        }
                    } else {
                        val newRequest = DataRequest(
                            this@DataProvider,
                            receiver,
                            areaBucket,
                            resolutionBucket,
                            type,
                            data
                        )
                        this@DataProvider.addRequestToQueue(key, newRequest);
                    }
                } finally {
                    decrementAwaitingQueueAddition()
                }
            }
        }
    }

    private fun createResolutionBucket(resolution: PointGeo): PointGeo {
        var x = log2(resolution.longitude)
        x = ceil(x / 2) * 2
        var y = log2(resolution.latitude)
        y = ceil(y / 2) * 2

        return PointGeo(2.0.pow(x), 2.0.pow(y))
    }

    private fun createAreaBucket(area: BoxGeo, resolution: PointGeo, type: OverlayTypes): BoxGeo {
        // Check if the area exists in a used bucket
        val resolutionKey = this.getResolutionKey(resolution);
        val importantBucket = this.usedBuckets[type] ?: throw Exception("Invalid bucket requested in DataProvider")
        var bucketOfKeys = importantBucket[resolutionKey]
        if (bucketOfKeys == null) {
            bucketOfKeys = LinkedList()
            importantBucket[resolutionKey] = bucketOfKeys
        }

        val areaTl = area.topLeft
        val areaBr = area.bottomRight
        for ( obj in bucketOfKeys ) {
            val tl = obj.topLeft
            val br = obj.bottomRight
            if ( tl.longitude <= areaTl.longitude && tl.latitude >= areaTl.latitude &&
                br.longitude >= areaBr.longitude && br.latitude <= areaBr.latitude ) {
                return obj
            }
        }

        // We need to create a new bucket
        val tlLongitude = areaTl.longitude - (area.width * areaBucketMultiplier)
        val tlLatitude = areaTl.latitude + (area.height * areaBucketMultiplier)
        val brLongitude = areaBr.longitude + (area.width * areaBucketMultiplier)
        val brLatitude = areaBr.latitude - (area.height * areaBucketMultiplier)
        val bucket = BoxGeo(PointGeo(this.forceValidLongitude(tlLongitude), this.forceValidLatitude(tlLatitude)),
            PointGeo(this.forceValidLongitude(brLongitude), this.forceValidLatitude(brLatitude)));
        bucketOfKeys.push(bucket);
        return bucket;
    }

    private fun forceValidLongitude(longitude: Double): Double {
        if ( longitude < -180 ) {
            return -180.0
        }
        if ( longitude > 180 ) {
            return 180.0
        }
        return longitude
    }

    private fun forceValidLatitude(latitude: Double): Double {
        if ( latitude < -90 ) {
            return -90.0
        }
        if ( latitude > 90 ) {
            return 90.0
        }
        return latitude
    }

    private fun getResolutionKey(resolution: PointGeo): String {
        return "${resolution.longitude}|${resolution.latitude}"
    }

    private fun getCacheKey(area: BoxGeo, resolution: PointGeo, type: OverlayTypes): String {
        val tl = area.topLeft
        val br = area.bottomRight
        return "${type}|${tl.longitude}|${tl.latitude}|${br.longitude}|${br.latitude}|${resolution.longitude}|${resolution.latitude}"
    }
}